src/fix_gemc.cpp

```diff
@@ -0,0 +1,514 @@
+  q_flag = atom->q_flag;
+
+  // pre compute scaled temperature
+  beta = 1.0/(force->boltz*box_temp);
+
+  // get domain dim
+  triclinic_flag = domain->triclinic;
+
+  // get domain dims
+  xlo = domain->boxlo[0];
+  xhi = domain->boxhi[0];
+  ylo = domain->boxlo[1];
+  yhi = domain->boxhi[1];
+  zlo = domain->boxlo[2];
+  zhi = domain->boxhi[2];
+
+  // get subdomain
+  if (triclinic_flag) {
+    sublo = domain->sublo_lamda;
+    subhi = domain->subhi_lamda;
+  } else {
+    sublo = domain->sublo;
+    subhi = domain->subhi;
+  }
+
+  // create unique group name for atoms to be excluded for particle exchange
+  // keeps temporarily deleted particles from being added in potential energy calc
+
+  // id from fix
+  auto group_id = std::string("FixGEMC:gemc_exclusion_group:") + id;
+  group->assign(group_id + " subtract all all");
+  exclusion_group = group->find(group_id);
+  if (exclusion_group == -1)
+    error->all(FLERR,"Could not find fix gemc exclusion group ID");
+  exclusion_group_bit = group->bitmask[exclusion_group];
+
+  // neighbor list exclusion setup
+  // turn off interactions between group all and the exclusion group
+
+  neighbor->modify_params(fmt::format("exclude group {} all",group_id));
+
+  // for exchange
+
+  maxbuf = (init_exchange() + BUFMIN) * BUFFACTOR;
+  memory->create(buf,maxbuf,"fix_gemc:buf");
+
+  groupbitall = 1 | groupbit;
+}
+
+/* ----------------------------------------------------------------------
+   attempt Monte Carlo translations, rotations, insertions, and deletions
+   done before exchange, borders, reneighbor
+   so that ghost atoms and neighbor lists will be correct
+
+   gcmc + extra volume step and modified insertion/deletion
+------------------------------------------------------------------------- */
+
+void FixGEMC::pre_exchange()
+{
+
+  // just return if should not be called on this timestep
+
+  if (next_reneighbor != update->ntimestep) return;
+
+  // get domain dims
+  xlo = domain->boxlo[0];
+  xhi = domain->boxhi[0];
+  ylo = domain->boxlo[1];
+  yhi = domain->boxhi[1];
+  zlo = domain->boxlo[2];
+  zhi = domain->boxhi[2];
+
+  // get subdomain
+  if (triclinic_flag) {
+    sublo = domain->sublo_lamda;
+    subhi = domain->subhi_lamda;
+  } else {
+    sublo = domain->sublo;
+    subhi = domain->subhi;
+  }
+
+  // three steps in GEMC:
+  // 1) translate particles within each box
+  // 2) exchange particles between boxes
+  // 3) change box volume + scale particle positions
+
+  // update next time to call
+  next_reneighbor = update->ntimestep + nevery;
+
+  // translation seems to be fully working
+  double imove;
+  update_gas_atoms_list();
+
+  energy_stored = energy_full();
+  int prev_step = 0;
+
+  if (energy_stored > 1E6)
+    error->one(FLERR,"fix gemc: Energy of old configuration big");
+
+
+  for (int i = 0; i < nmoves; i++) {
+    imove = random_universe->uniform();
+
+    if (fabs(energy_full()) > 1E8) {
+      printf("step: %i; prev_step: %i\n", i, prev_step);
```

Comment:
Please *never* use printf() always output to "screen" and/or "logfile". This is most conveniently done using `utils::logmesg()`.